### ISSUE TITLE
[YSQL] Consistency issue: the row both exists and doesn't exist at the same time

### ISSUE THREAD

**Issue by slava-vishnyakov** [yugabyte/yugabyte-db#10105]
Jira Link: [DB-4760](https://yugabyte.atlassian.net/browse/DB-4760)
I have a weird issue where a row both exists and doesn't exist at the same time, it has a unique index on it, but it can be added 2 times.

```sql
yugabyte=# select * from kv; 
   k   | v
-------+---
 count | 0
(1 row)

yugabyte=# select * from kv where k = 'count'; 
 k | v
---+---
(0 rows)  <--- what? where is it?

yugabyte=# select count(*) from kv;
 count
-------
     1
(1 row)
```

So, if you call `select * from kv where k = 'count'`, then the row does not exist, but if you call `select * from kv`, then it exists. 

At the same time, despite the fact that I have a unique index on `k`, I can add it again, but only once:

```sql
yugabyte=# insert into kv (k,v) values ('count', 0);
INSERT 0 1

yugabyte=# insert into kv (k,v) values ('count', 0);
ERROR:  duplicate key value violates unique constraint "kv_k_idx"

yugabyte=# select k, count(v) from kv group by 1;
   k   | count
-------+-------
 count |     2
(1 row)
```

The bug is a bit hard to reproduce, as it can't be reproduced using `psql`, but rather you need to do a Python call using `orator` library:

```python
from orator import DatabaseManager

if __name__ == '__main__':
    config = {
        'pgsql': {
            'driver': 'pgsql',
            'host': '10.1.0.2',
            'database': 'yugabyte',
            'user': 'yugabyte',
            'port': 5433,
            'password': 'yugabyte'
        }
    }
    db = DatabaseManager(config)

    db.update('''
        drop table if exists kv;
        create table kv (k text, v int); 
        insert into kv (k, v) values ('count', 42); 
        create unique index on kv (k);
    ''')

    print('the table contains:', db.select('select * from kv'))
    print(db.select("select v from kv where k = 'count'"), '<-- this should be [42]')
    print(db.select("select version()")[0][0])
    print('can we add another one?')
    db.update('''insert into kv (k, v) values ('count', 42);''')
    print('the table now contains:', db.select('select * from kv'), 'despite the unique index')
```

Output:

```console
the table contains: [['count', 42]]
[] <-- this should be [42]
PostgreSQL 11.2-YB-2.9.0.0-b0 on x86_64-pc-linux-gnu, compiled by gcc (Homebrew gcc 5.5.0_4) 5.5.0, 64-bit
can we add another one?
the table now contains: [['count', 42], ['count', 42]] despite the unique index
```

Just to note, if you connect to Postgres with the same code, it doesn't have this issue.

I guess the problem is with this call:

```python
db.update('''
    drop table if exists kv;
    create table kv (k text, v int); 
    insert into kv (k, v) values ('cnt', 42); 
    create unique index on kv (k);
''')
```

...but then, again, Postgresql doesn't have any problems with it.

If I change the order to this, it works correctly:

```python
db.update('''
    drop table if exists kv;
    create table kv (k text, v int); 
    create unique index on kv (k);
    insert into kv (k, v) values ('cnt', 42); 
''')
```

**Comment by jasonyb** [comment]
Reproduced on recent master a1edbd289e6, `PostgreSQL 11.2-YB-2.17.2.0-b0 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9), 64-bit`.

So far, I notice sst_dump doesn't show the record right after creating the index (added a sleep there).

**Comment by jasonyb** [comment]
The issue is index corruption when CREATE INDEX is used in a transaction block.  It is not very surprising.  Marking as duplicate of #5240.  See also #6240 and #1404.

The statements in a `db.update` get bundled into a transaction.  DDLs within transaction blocks are not officially supported.  What's likely happening is that the CREATE INDEX does not see the previously inserted row since it is still provisional (not committed), so the index is empty.  Also, note that the CREATE INDEX in transaction block is defaulted to be NONCONCURRENT, which is unsafe during concurrent operations (there are none in this test case).

Running the python script with `create index` in a separate `db.update` (and also modifying it to say `create index nonconcurrently` to preserve the same behavior as when it was in the same `db.update`), the issue is resolved.  Running the statements in ysqlsh using explicit BEGIN/COMMIT transaction blocks for the contents of `db.update` also reproduces the issue.

```py
from orator import DatabaseManager                                                                                                                                                                                                                                                                                                                                           
import argparse                                                                                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                                             
parser = argparse.ArgumentParser()                                                                                                                                                                                                                                                                                                                                           
parser.add_argument('host', type=str, help='host')                                                                                                                                                                                                                                                                                                                           
parser.add_argument('case', type=str, choices=('good', 'bad'), help='test case')                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                                             
if __name__ == '__main__':                                                                                                                                                                                                                                                                                                                                                   
    args = parser.parse_args()                                                                                                                                                                                                                                                                                                                                               
    config = {                                                                                                                                                                                                                                                                                                                                                               
        'pgsql': {                                                                                                                                                                                                                                                                                                                                                           
            'driver': 'pgsql',                                                                                                                                                                                                                                                                                                                                               
            'host': args.host,                                                                                                                                                                                                                                                                                                                                               
            'database': 'yugabyte',                                                                                                                                                                                                                                                                                                                                          
            'user': 'yugabyte',                                                                                                                                                                                                                                                                                                                                              
            'port': 5433,                                                                                                                                                                                                                                                                                                                                                    
            'password': 'yugabyte'                                                                                                                                                                                                                                                                                                                                           
        }                                                                                                                                                                                                                                                                                                                                                                    
    }                                                                                                                                                                                                                                                                                                                                                                        
    db = DatabaseManager(config)                                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                                             
    if args.case == 'good':                                                                                                                                                                                                                                                                                                                                                  
        db.update('''                                                                                                                                                                                                                                                                                                                                                        
            drop table if exists kv;                                                                                                                                                                                                                                                                                                                                         
            create table kv (k text, v int);                                                                                                                                                                                                                                                                                                                                 
            insert into kv (k, v) values ('count', 42);                                                                                                                                                                                                                                                                                                                      
        ''')                                                                                                                                                                                                                                                                                                                                                                 
        db.update('''                                                                                                                                                                                                                                                                                                                                                        
            create unique index nonconcurrently on kv (k);                                                                                                                                                                                                                                                                                                                   
        ''')                                                                                                                                                                                                                                                                                                                                                                 
    else:                                                                                                                                                                                                                                                                                                                                                                    
        db.update('''                                                                                                                                                                                                                                                                                                                                                        
            drop table if exists kv;                                                                                                                                                                                                                                                                                                                                         
            create table kv (k text, v int);                                                                                                                                                                                                                                                                                                                                 
            insert into kv (k, v) values ('count', 42);                                                                                                                                                                                                                                                                                                                      
            create unique index nonconcurrently on kv (k);                                                                                                                                                                                                                                                                                                                   
        ''')                                                                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                             
    print('the table contains:', db.select('select * from kv'))                                                                                                                                                                                                                                                                                                              
    print(db.select("select v from kv where k = 'count'"), '<-- this should be [42]')                                                                                                                                                                                                                                                                                        
    print(db.select("select version()")[0][0])                                                                                                                                                                                                                                                                                                                               
    print('can we add another one?')                                                                                                                                                                                                                                                                                                                                         
    try:                                                                                                                                                                                                                                                                                                                                                                     
        db.update('''insert into kv (k, v) values ('count', 42);''')                                                                                                                                                                                                                                                                                                         
        print('the table now contains:', db.select('select * from kv'), 'despite the unique index')                                                                                                                                                                                                                                                                          
    except Exception:                                                                                                                                                                                                                                                                                                                                                        
        print('no')
```
